api/envoy/config/listener/v3/listener.proto

```diff
@@ -275,4 +280,26 @@ message Listener {
   // :ref:`use_original_dst <envoy_api_field_config.listener.v3.Listener.use_original_dst>`
   // to true. Default is true.
   google.protobuf.BoolValue bind_to_port = 26;
+
+  // The exclusive listener type and the corresponding config.
+  // TODO(lambdai): https://github.com/envoyproxy/envoy/issues/15372
+  // Will create and add TcpListenerConfig. Will add UdpListenerConfig and ApiListener.
+  // [#not-implemented-hide:]
+  oneof listener_specifier {
+    // Used to represent an internal listener which does not listen on OSI L4 address but can be used by the
+    // :ref:`envoy cluster <envoy_v3_api_msg_config.cluster.v3.Cluster>` to create a user space connection to.
+    // The internal listener acts as a tcp listener. It supports listener filters and network filter chains.
+    // The internal listener require :ref:`address <envoy_api_field_config.listener.v3.Listener.address>` has
+    // field `envoy_internal_address`.
+    //
+    // There are some limitations are derived from the implementation. The known limitations include
+    //
+    // * :ref:`ConnectionBalanceConfig <envoy_api_msg_config.listener.v3.Listener.ConnectionBalanceConfig>` is not
+    //   allowed because both cluster connection and listener connection must be owned by the same dispatcher.
+    // * :ref:`tcp_backlog_size <envoy_api_field_config.listener.v3.Listener.tcp_backlog_size>`
+    // * :ref:`freebind <envoy_api_field_config.listener.v3.Listener.freebind>`
+    // * :ref:`transparent <envoy_api_field_config.listener.v3.Listener.transparent>`
```

Comment:
Do internal listeners support passing additional metadata from source? This would be interesting to be able to propagate filter metadata or filter state without the wire encoding.

Reply:
That's a good question.
Let me complete the doc and share it with you. It's easy to implement but it's hard to utilize such ability. The api will be weird if the proportion only work with internal connection